File: 0015/two_pointers.py
from typing import *


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        m = len(nums)
        ret = []
        for i in range(m):
            if i-1 >= 0 and nums[i-1] == nums[i]:
                continue
            rest = 0 - nums[i]
            left = i + 1
            right = m - 1
            while left < right:
                total = nums[left] + nums[right]
                if total > rest:
                    right -= 1
                elif total < rest:
                    left += 1
                else:
                    ret.append((nums[i], nums[left], nums[right]))
                    while left < right and nums[left] == nums[left+1]:
                        left += 1
                    while left < right and nums[right-1] == nums[right]:
                        right -= 1
                    left += 1
                    right -= 1
        return ret


if __name__ == '__main__':
    sol = Solution()

    # -4, -1, -1, 0, 1, 2
    nums = [-1, 0, 1, 2, -1, -4]
    expected = [(-1, -1, 2),(-1, 0, 1)]
    assert sol.threeSum(nums) == expected

    nums = []
    expected = []
    assert sol.threeSum(nums) == expected

File: 0015/two_pointers.cpp
#include <gtest/gtest.h>

#include <algorithm>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<vector<int>> threeSum(vector<int>& nums) {
    vector<vector<int>> ret;
    sort(nums.begin(), nums.end());
    size_t j, k;
    for (size_t i = 0; i < nums.size() - 2; i++) {
      if (0 < i && nums[i - 1] == nums[i]) { continue; }
      j = i + 1, k = nums.size() - 1;
      while (j < k) {
        int sum = nums[i] + nums[j] + nums[k];
        if (sum > 0) {
          k--;
        } else if (sum < 0) {
          j++;
        } else {
          ret.push_back({nums[i], nums[j++], nums[k--]});
          while (j < k && nums[k] == nums[k + 1]) { k--; }
          while (j < k && nums[j - 1] == nums[j]) { j++; }
        }
      }
    }
    return ret;
  }
};

int main() {
  Solution sol;
  vector<int> nums;
  vector<vector<int>> expected;

  nums = {-1, 0, 1, 2, -1, -4}, expected = {{-1, -1, 2}, {-1, 0, 1}};
  EXPECT_EQ(sol.threeSum(nums), expected);

  nums = {0, 0, 0, 0}, expected = {{0, 0, 0}};
  EXPECT_EQ(sol.threeSum(nums), expected);

  exit(EXIT_SUCCESS);
}
